latex/Inhalt/design.tex

```diff
@@ -1 +1,52 @@
+
+\section{Emulator}
+
+\subsection{Zentrale Struktur}
+
+Den Kern des Emulators bildet eine Struktur, welche zuständig für die Ausführung der Maschinencode-Programme ist. Diese Struktur gruppiert alle notwendigen Komponenten eines Intel 8080 Systems. Der Aufbau der Struktur ist in \cref{core} illustriert.
+
+Diese Komponenten wurden in \cref{chap:prereqs} bereits erklärt: \rust{pc} und \rust{sp} sind 2 16-Bit-Zahlen, die den Program Counter und den Stack Pointer repräsentieren. \rust{ram} ist der Arbeitsspeicher und \rust{reg} simuliert die Register (inklusive Flags und Akkumulator).
+Die Ports für I/O-Geräte werden durch 2 Arrays mit jeweils 256 Elementen repräsentiert.
+Darauf folgt eine Boolean, die aussagt ob der Emulator am Laufen ist und die Boolean die anzeigt ob Interrupts erlaubt sind.
+
+\begin{listing}[h]
+\label[code]{core}
+\begin{minted}{rust}
+    struct Emulator {
+        pc: u16,
+        sp: u16,
+        ram: RAM,
+        reg: RegisterArray,
+        input_devices: [InputDevice; 256],
+        output_devices: [OutputDevice; 256],
+        running: bool,
+        interrupts_enabled: bool
+    }
+\end{minted}
+\end{listing}
+
+\subsection{Modularität}
+
+Der Intel 8080 ist lediglich die CPU, RAM und I/O-Geräte arbeiten prinzipiell unabhängig. Diese müssen zwar die entsprechende Schnittstelle haben um angeschlossen werden zu können, aber können beliebig implementiert sein. Unsere Implementierung ermöglicht verschiedene Implementierungen für RAM und Input/Output-Devices zu haben. Es handelt sich bei diesen Typen jedoch nicht um Interfaces, da Rust diese nicht unterstützt. Wie genau das in Rust umgesetzt ist, wird in \cref{chap:impl} erläutert. Prinzipiell funktioniert es aber wie bei Interfaces, die notwendigen Methoden sind explizit deklariert, aber die Implementierungen sind beliebig.
+
+\subsection{Ausführung}
+
+Die \rust{Emulator::execute_next()} Methode führt die Instruktion an der Addresse im PC aus. Der Opcode wird über ein enormes \rust{match}-Statement auf die entsprechende funktion delegiert, die den Opcode entsprechend ausführt.
+
+Der Rückgabetyp der Methode ist \rust{Result<(), &str>}, dadurch können entsprechende Errormeldungen nach außen propagiert werden. Dies ist wünschenswert, damit auf dem Frontend entsprechende Errormeldungen angezeigt werden können, um dem Benutzer den Entwicklungsprozess zu erleichtern.
+
+\subsubsection{Instruktionen}
+
+Um zu großen Dateien vorzubeugen, sind die Implementierungen der Instruktionen aufgeteilt in verschiedene Module. Sie sind logisch gruppiert in Arithmetik, Kontrollfluss, Logik, Speicherzugriff, Verschiebung und Speziell.
+Obwohl die Funktionen in unterschiedlichen Dateien/Modulen deklariert sind, sind sie Methoden der \rust{Emulator}-Struktur.
+Die verschiedenen Funktionen werden dann im Code von \rust{Emulator::execute_next()} aufgerufen.
+Auch diese Funktionen geben häufig \rust{Result}s zurück, falls die Ausführung theoretisch in einem Fehler resultieren könnte.
```

Comment:
"sofern die Ausführung ~theoretisch~ in einem Fehler resultieren ~könnte~ _kann_"